### PR TITLE
Adding initial GPSSetup project

### PR DESCRIPTION
**Pull request by pburma**
GPSSetup is a client builder example of how to convert XY image coordinate points to GPS lat/lng points for the ConvertCoordinates activity pattern.

=== Instructions ===

- Run the GPSSetup client. 
- Click anywhere on the map to add camera markers
- Use the GPS button on the table to open the Calibration page
- Click Load Image to load a local file into the html canvas
- Move the 4 green squares around to indicate field of view for a camera
- Move the 4 GPS points located on the map to the right around to locate physical location
- Click Submit to save

The GPS and XY coordinates are saved to the cameras type. Once saved to the cameras type a CachedEnriched event can be used to load those values into each memory and then passed into a ConvertCoordinated activity pattern as a downstream event. An example application in the App Builder will look like this:

EventStream(objrec payload) --> SplitBy CameraId --> CachedEnrich cameras --> Transformation --> ConvertCoordinates --> TrackMotion

The Transformation can be very tricky so an example procedure has been included in this project called transformCameraEvents that can be applied to the ConvertCoordinates calibrationKey parameters using the value event.calibrationCoordinates. 

=== Typical Uses ===

To show the results of object recognition data on a map. Object recognition systems will return 4 point bounding box results that indicate the location of an object inside its field of view, which is just an image or a frame from a video. To show where in the real-world that object is located a user can use these web tools to calibrate the camera mapping settings in a simple and easy way. 

Zoning:
It is also possible to use similar UI components to create a Zone (area of interest) in the image. The GPSSetup client can be modified to save and load the 4 points from an image and not include any map points. An object property has to be added to store the Zone for each camera. There is no "Zone" activity pattern so a procedure would need to be written that looks to see if a bounding box or centroid provided by the computer vision results is located inside your zone of interest. The procedure pointInPolygon was added to this project as an example of how to make that calculation.

=== Caveats and Other Notes ===

If the Vantiq extension source is being used it has the ability to upload images periodically or on-demand using a remote query. The Image Load feature can then be replaced by loading the most recent image from the camera of that same name. 

If the Vantiq extension source is not being used it may be possible to load an image from a recent payload from the CVE system otherwise a user will have to manually upload an image. A property was added to the type called "img" which can be used to reference an image to load here but remember if the camera is ever moved a new upload must be made.

### CODE REVIEW

**Comment by pburma** [comment]
image

image

File: GPSSetup/clients/GPSSetup/controllers/client/customCode.js [hunk 1]
```diff
@@ -0,0 +1,357 @@
+// All the drawing stuff for the camera setup
+//
+//
+
+function Shape(x, y, w, h, fill) {
+    console.log("CALLED: Shape");
+    // This is a very simple and unsafe constructor. All we're doing is checking if the values exist.
+    // "x || 0" just means "if there is a value for x, use that. Otherwise use 0."
+    // But we aren't checking anything else! We could put "Lalala" for the value of x
```
Comment (namirfa) [inline GPSSetup/clients/GPSSetup/controllers/client/customCode.js:9]: Is it worth checking that x/y/w/h are all integers, and that fill is a valid hex color string?

File: GPSSetup/clients/GPSSetup/controllers/client/customCode.js [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+
+
+function CanvasState(canvas, backgroundimage, page) {
+    console.log("CALLED: CanvasState");
+    // **** First some setup! ****
+
+    this.page = page;
+
+    this.base_image = backgroundimage;
+    //this.base_image = canvas.toDataURL();
```
Comment (namirfa) [inline GPSSetup/clients/GPSSetup/controllers/client/customCode.js:42]: Remove commented code

File: GPSSetup/clients/GPSSetup/controllers/client/customCode.js [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+
+    this.canvas = canvas;
+    this.width = canvas.width;
+    this.height = canvas.height;
+    this.ctx = canvas.getContext('2d');
+
+    // This complicates things a little but but fixes mouse co-ordinate problems
```
Comment (namirfa) [inline GPSSetup/clients/GPSSetup/controllers/client/customCode.js]:
```suggestion
    // This complicates things a little but fixes mouse coordinate problems
```

File: GPSSetup/clients/GPSSetup/controllers/client/customCode.js [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+    // Right here "this" means the CanvasState. But we are making events on the Canvas itself,
+    // and when the events are fired on the canvas the variable "this" is going to mean the canvas!
+    // Since we still want to use this particular CanvasState in the events we have to save a reference to it.
+    // This is our reference!
+    var myState = this;
+
+    var type = myState.page.data.zone;
+
+
+    //fixes a problem where double clicking causes text to get selected on the canvas
```
Comment (namirfa) [inline GPSSetup/clients/GPSSetup/controllers/client/customCode.js]:
```suggestion
    // Fixes a problem where double clicking causes text to get selected on the canvas
```

File: GPSSetup/clients/GPSSetup/controllers/client/customCode.js [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+    canvas.addEventListener('selectstart', function(e) { e.preventDefault(); return false; }, false);
+    // Up, down, and move are for dragging
+    canvas.addEventListener('mousedown', function(e) {
+        // if screen scrolls we still need to be able to move the markers reliably
+        var scroll = $("#_ScrollingFrame_").scrollTop();
+        console.log(scroll);
```
Comment (namirfa) [inline GPSSetup/clients/GPSSetup/controllers/client/customCode.js:92]: What is this logging/is it worth logging? Just wondering if that was left behind from debugging.

File: GPSSetup/clients/GPSSetup/controllers/client/customCode.js [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+        if (myState.dragging){
+
+            var shapes = myState.shapes;
+            var mouse = myState.getMouse(e);
+            // We don't want to drag the object by its top-left corner, we want to drag it
+            // from where we clicked. Thats why we saved the offset and use it here
+            myState.selection.x = mouse.x - myState.dragoffx;
+            myState.selection.y = mouse.y - myState.dragoffy;  
+
+            // myState.page.data.camera.zoneConfiguration = [{"zone":"zone1"}]
```
Comment (namirfa) [inline GPSSetup/clients/GPSSetup/controllers/client/customCode.js:129]: Remove commented code

File: GPSSetup/clients/GPSSetup/controllers/client/customCode.js [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+                    console.log(myState);
+
+
+                    // case 0-3 is the gps zone, case 4-6 is dwell zone
+                    switch(i + type) {
+                        case 0:
+                            myState.page.data.camera.imageTrack.topLeftImage = {"x":myState.selection.x,"y":myState.selection.y};
+                            console.log("TOPLEFTIMAGE");
+                            console.log(myState.page);
+                            console.log(myState.page.topLeftImage);
```
Comment (namirfa) [inline GPSSetup/clients/GPSSetup/controllers/client/customCode.js:145]: I'm guessing these log statements are left behind from debugging. Probably better to delete them, unless we want to add logs for each case?

File: GPSSetup/clients/GPSSetup/controllers/client/customCode.js [hunk 8]
```diff
@@ -0,0 +1,357 @@
...
+            // We don't want to drag the object by its top-left corner, we want to drag it
+            // from where we clicked. Thats why we saved the offset and use it here
+            myState.selection.x = mouse.x - myState.dragoffx;
+            myState.selection.y = mouse.y - myState.dragoffy;  
+
+            // myState.page.data.camera.zoneConfiguration = [{"zone":"zone1"}]
+
+            for (var i = shapes.length-1; i >= 0; i--) {
+                if ( (shapes[i].x === myState.selection.x) && (shapes[i].y === myState.selection.y) ) {
+                    console.log("You are moving shape: "  + i );
```
Comment (namirfa) [inline GPSSetup/clients/GPSSetup/controllers/client/customCode.js]:
```suggestion
                    console.log("You are moving shape: "  + i);
```

File: GPSSetup/clients/GPSSetup/controllers/client/customCode.js [hunk 9]
```diff
@@ -0,0 +1,357 @@
...
+                    console.log(type);
+                    console.log( i + type);
```
Comment (namirfa) [inline GPSSetup/clients/GPSSetup/controllers/client/customCode.js:135]:
```suggestion
                    console.log(i + type);
```

File: GPSSetup/clients/GPSSetup/controllers/client/customCode.js [hunk 10]
```diff
@@ -0,0 +1,357 @@
...
+                            myState.page.data.camera.zoneConfiguration[0].bottomleft = {"x":myState.selection.x,"y":myState.selection.y};
+                            break;
+                        case 6:
+                            myState.page.data.camera.zoneConfiguration[0].bottomright = {"x":myState.selection.x,"y":myState.selection.y};
+                            break;
+                        case 7:
+                            myState.page.data.camera.zoneConfiguration[0].topright = {"x":myState.selection.x,"y":myState.selection.y};
+                            break;
+                        default:
+                            // code block
```
Comment (namirfa) [inline GPSSetup/clients/GPSSetup/controllers/client/customCode.js:171]: Shouldn't there be a default case here, or is this just ignored? Is it worth logging an error or something of the sort?

File: GPSSetup/clients/GPSSetup/controllers/client/customCode.js [hunk 11]
```diff
@@ -0,0 +1,357 @@
...
+                    }
+
+                    console.log(myState.page.data.camera.zoneConfiguration);
+
+
```
Comment (namirfa) [inline GPSSetup/clients/GPSSetup/controllers/client/customCode.js:176]: Probably don't need empty lines surrounding this, or at least not the two empty lines after the log.

File: GPSSetup/clients/GPSSetup/controllers/client/customCode.js [hunk 12]
```diff
@@ -0,0 +1,357 @@
...
+            }
+
+            console.log(myState.selection.x + " " + myState.selection.y);
+            console.log(myState);
+            myState.valid = false; // Something's dragging so we must redraw
+        }
+    }, true);
+    canvas.addEventListener('mouseup', function(e) {
+        //console.log("YOU LET GO");
+        // console.log(this);
```
Comment (namirfa) [inline GPSSetup/clients/GPSSetup/controllers/client/customCode.js:187]: Remove commented code

File: GPSSetup/clients/GPSSetup/controllers/client/customCode.js [hunk 13]
```diff
@@ -0,0 +1,357 @@
...
+    this.interval = 30;
+    setInterval(function() { myState.draw(); }, myState.interval);
+}
+
+CanvasState.prototype.changeBaseImage = function(image) {
+
+    this.base_image = image;
+
+
+
```
Comment (namirfa) [inline GPSSetup/clients/GPSSetup/controllers/client/customCode.js:216]: Don't need so many empty lines

File: GPSSetup/clients/GPSSetup/controllers/client/customCode.js [hunk 14]
```diff
@@ -0,0 +1,357 @@
...
+    this.shapes.push(shape);
+    this.valid = false;
+
+};
+
+
+CanvasState.prototype.moveShape = function(shapenumber, x,y) {
+
+    console.log("CALLED: CanvasState.prototype.addShape");
+    // this.shapes.push(shape);
```
Comment (namirfa) [inline GPSSetup/clients/GPSSetup/controllers/client/customCode.js:231]: Remove commented code

File: GPSSetup/clients/GPSSetup/controllers/client/customCode.js [hunk 15]
```diff
@@ -0,0 +1,357 @@
...
+
+    this.shapes[shapenumber].x = x;
+    this.shapes[shapenumber].y = y;
+
```
Comment (namirfa) [inline GPSSetup/clients/GPSSetup/controllers/client/customCode.js:235]: Don't need two empty lines, one should be sufficient

File: GPSSetup/clients/GPSSetup/controllers/client/customCode.js [hunk 16]
```diff
@@ -0,0 +1,357 @@
...
+};
+
+CanvasState.prototype.addShape = function(shape) {
+
+    console.log("CALLED: CanvasState.prototype.addShape");
+    this.shapes.push(shape);
+    this.valid = false;
+
+};
+
```
Comment (namirfa) [inline GPSSetup/clients/GPSSetup/controllers/client/customCode.js:226]: Only need one line between method defs

File: GPSSetup/clients/GPSSetup/controllers/client/customCode.js [hunk 17]
```diff
@@ -0,0 +1,357 @@
...
+    // this.shapes.push(shape);
+
+    this.shapes[shapenumber].x = x;
+    this.shapes[shapenumber].y = y;
+
+
+    this.valid = false;
+
+};
+
```
Comment (namirfa) [inline GPSSetup/clients/GPSSetup/controllers/client/customCode.js:240]: Only need one line between method defs

File: GPSSetup/clients/GPSSetup/controllers/client/customCode.js [hunk 18]
```diff
@@ -0,0 +1,357 @@
...
+CanvasState.prototype.clear = function() {
+    console.log("CALLED: CanvasState.prototype.clear");
+    this.ctx.clearRect(0, 0, this.width, this.height);
+};
+
+// While draw is called as often as the INTERVAL variable demands,
+// It only ever does something if the canvas gets invalidated by our code
+CanvasState.prototype.draw = function() {
+    // console.log("CALLED: CanvasState.prototype.draw");
```
Comment (namirfa) [inline GPSSetup/clients/GPSSetup/controllers/client/customCode.js:250]: Probably best to either uncomment this log line, or delete all of the initial "CALLED" logs for each method.

File: GPSSetup/clients/GPSSetup/controllers/client/customCode.js [hunk 19]
```diff
@@ -0,0 +1,357 @@
...
+};
+
+
+CanvasState.prototype.clear = function() {
+    console.log("CALLED: CanvasState.prototype.clear");
+    this.ctx.clearRect(0, 0, this.width, this.height);
+};
+
+// While draw is called as often as the INTERVAL variable demands,
+// It only ever does something if the canvas gets invalidated by our code
```
Comment (namirfa) [inline GPSSetup/clients/GPSSetup/controllers/client/customCode.js]:
```suggestion
// it only ever does something if the canvas gets invalidated by our code
```

File: GPSSetup/clients/GPSSetup/controllers/client/customCode.js [hunk 20]
```diff
@@ -0,0 +1,357 @@
...
+
+                shapes[i].draw(ctx);
+
+                if(i === 0)
+                    ctx.moveTo(shape.x, shape.y);
+                else if (i < 4)
+                    ctx.lineTo(shape.x, shape.y);
+            }
+            ctx.closePath();
+            // ctx.fillStyle = '#f00';
```
Comment (namirfa) [inline GPSSetup/clients/GPSSetup/controllers/client/customCode.js:288]: Remove commented code

File: GPSSetup/clients/GPSSetup/controllers/client/customCode.js [hunk 21]
```diff
@@ -0,0 +1,357 @@
...
+            ctx.fillStyle = "rgba(193, 66, 66, 0.4)";
+            //ctx.globalAlpha = 0.5;
```
Comment (namirfa) [inline GPSSetup/clients/GPSSetup/controllers/client/customCode.js:290]: Remove commented code

File: GPSSetup/clients/GPSSetup/controllers/client/customCode.js [hunk 22]
```diff
@@ -0,0 +1,357 @@
...
+
+                shapes[i].draw(ctx);
+
+                if(i === 4)
+                    ctx.moveTo(shape.x, shape.y);
+                else if (i < 8)
+                    ctx.lineTo(shape.x, shape.y);
+            }
+            ctx.closePath();
+            // ctx.fillStyle = '#f00';
```
Comment (namirfa) [inline GPSSetup/clients/GPSSetup/controllers/client/customCode.js:311]: Remove commented code

File: GPSSetup/clients/GPSSetup/controllers/client/customCode.js [hunk 23]
```diff
@@ -0,0 +1,357 @@
...
+            ctx.fillStyle = "rgba(66, 66, 193, 0.7)";
+            //ctx.globalAlpha = 0.5;
```
Comment (namirfa) [inline GPSSetup/clients/GPSSetup/controllers/client/customCode.js:313]: Remove commented code

File: GPSSetup/clients/GPSSetup/controllers/client/customCode.js [hunk 24]
```diff
@@ -0,0 +1,357 @@
...
+            ctx.fill();
```
Comment (namirfa) [inline GPSSetup/clients/GPSSetup/controllers/client/customCode.js:314]: The code for `l >= 4` and `l >= 8` is almost identical. Probably could be consolidated in a helper method.

Also, is `l` ever anything but 4 or 8? What happens if it's less than 4?

File: GPSSetup/clients/GPSSetup/controllers/client/customCode.js [hunk 25]
```diff
@@ -0,0 +1,357 @@
...
+CanvasState.prototype.getMouse = function(e) {
+    console.log("CALLED: CanvasState.prototype.getMouse");
+    var element = this.canvas, offsetX = 0, offsetY = 0, mx, my;
+
+    // Compute the total offset
+    if (element.offsetParent !== undefined) {
+        do {
+            offsetX += element.offsetLeft;
+            offsetY += element.offsetTop;
+        } while ((element = element.offsetParent));
```
Comment (namirfa) [inline GPSSetup/clients/GPSSetup/controllers/client/customCode.js:344]: Any reason why this can't simply be a while loop? Doesn't look like the "do" block affects the outcome of the while condition.

File: GPSSetup/clients/GPSSetup/controllers/page/CameraGPS/onEnd.js [hunk 26]
```diff
@@ -0,0 +1,12 @@
+    // clear image
```
Comment (namirfa) [inline GPSSetup/clients/GPSSetup/controllers/page/CameraGPS/onEnd.js:1]: Could be a Vantiq Client thing, but any reason why this code is indented?

File: GPSSetup/clients/GPSSetup/controllers/page/CameraGPS/onStart.js [hunk 27]
```diff
@@ -0,0 +1,197 @@
+    var page = this;
```
Comment (namirfa) [inline GPSSetup/clients/GPSSetup/controllers/page/CameraGPS/onStart.js:1]: Ditto - any reason why this code is indented?

File: GPSSetup/clients/GPSSetup/controllers/page/CameraGPS/onStart.js [hunk 28]
```diff
@@ -0,0 +1,197 @@
...
+    console.log(parameters);
+
+    let lat = page.data.camera.geo.coordinates[1];
+    let lng = page.data.camera.geo.coordinates[0];
+    var place = {lat: lat, lng: lng};
+    map.setMapTypeId('satellite');
+    map.setCenter(place);
+    map.setZoom(18);    
+
+    //var canvas = $("#callibration-canvas").get(0);
```
Comment (namirfa) [inline GPSSetup/clients/GPSSetup/controllers/page/CameraGPS/onStart.js:24]: Remove commented code

File: GPSSetup/clients/GPSSetup/controllers/page/CameraGPS/onStart.js [hunk 29]
```diff
@@ -0,0 +1,197 @@
...
+    var imageLoader = document.getElementById('imageLoader');
+    imageLoader.addEventListener('change', handleImage, false);
+
+    // setup the map gps markerss
+    client.data.camGeoSetup.initializeToDefaultValues();
+    var latlngbl, latlngbr, latlngtl, latlngtr;
+    console.log(page.data.camera.gpsTrack);
+
+    if (jQuery.isEmptyObject(page.data.camera.gpsTrack) === false){
+        //if (page.data.camera.gpsTrack.length > 1 ){
```
Comment (namirfa) [inline GPSSetup/clients/GPSSetup/controllers/page/CameraGPS/onStart.js:39]: Remove commented code

File: GPSSetup/clients/GPSSetup/controllers/page/CameraGPS/onStart.js [hunk 30]
```diff
@@ -0,0 +1,197 @@
...
+        addMarker(latlngbl, "bottomLeftGeo");
+
+        latlngbr = {lat:page.data.camera.geo.coordinates[1] - 0.0001, lng:page.data.camera.geo.coordinates[0] + 0.0001};
+        addMarker(latlngbr, "bottomRightGeo");
+
+        latlngtl = {lat:page.data.camera.geo.coordinates[1] + 0.0001, lng:page.data.camera.geo.coordinates[0] - 0.0001};
+        addMarker(latlngtl, "topLeftGeo");
+
+        latlngtr = {lat:page.data.camera.geo.coordinates[1] + 0.0001, lng:page.data.camera.geo.coordinates[0] + 0.0001};
+        addMarker(latlngtr, "topRightGeo");
```
Comment (namirfa) [inline GPSSetup/clients/GPSSetup/controllers/page/CameraGPS/onStart.js:66]: Looks like the only difference between the if/else blocks is how the `lat___` variables are defined. Could simplify by  setting them once, and conditionally changing them if necessary.

Then, outside the if block, you only need to call `addMarker` 4 times.

Something like:

```
var latlngbl = page.data.camera.gpsTrack.bottomLeftGeo;
var latlngbr = page.data.camera.gpsTrack.bottomRightGeo;
var latlngtl = page.data.camera.gpsTrack.topLeftGeo;
var latlngtr = page.data.camera.gpsTrack.topRightGeo;

if (jQuery.isEmptyObject(page.data.camera.gpsTrack)) {
    latlngbl = {lat:page.data.camera.geo.coordinates[1] - 0.0001, lng:page.data.camera.geo.coordinates[0] - 0.0001};
    latlngbr = {lat:page.data.camera.geo.coordinates[1] - 0.0001, lng:page.data.camera.geo.coordinates[0] + 0.0001};
    latlngtl = {lat:page.data.camera.geo.coordinates[1] + 0.0001, lng:page.data.camera.geo.coordinates[0] - 0.0001};
    latlngtr = {lat:page.data.camera.geo.coordinates[1] + 0.0001, lng:page.data.camera.geo.coordinates[0] + 0.0001};
}

addMarker(latlngbl, "bottomLeftGeo");
addMarker(latlngbr, "bottomRightGeo");
addMarker(latlngtl, "topLeftGeo");
addMarker(latlngtr, "topRightGeo");
```

File: GPSSetup/clients/GPSSetup/controllers/page/CameraGPS/onStart.js [hunk 31]
```diff
@@ -0,0 +1,197 @@
...
+            // set the geoJSON for the object
+            page.data.camera.gpsTrack[this.title] = {
+                "lng": location.lng(),
+                "lat": location.lat()
+            };
+
+        }); 
+
+        page.data.markers.push(marker);
+        //marker.setMap(map);
```
Comment (namirfa) [inline GPSSetup/clients/GPSSetup/controllers/page/CameraGPS/onStart.js:94]: Remove commented code

File: GPSSetup/clients/GPSSetup/controllers/page/CameraGPS/onStart.js [hunk 32]
```diff
@@ -0,0 +1,197 @@
...
+            var x = page.data.camera.imageTrack.topRightImage.x;
+            var y = page.data.camera.imageTrack.topRightImage.y;
+            page.data.CanvasState.moveShape(3, x,y);
+
+        } else {
+            page.data.CanvasState.moveShape(3, 80,20);
+        }
+
+
+    }
```
Comment (namirfa) [inline GPSSetup/clients/GPSSetup/controllers/page/CameraGPS/onStart.js:138]: Remove unnecessary new lines between if/else, and inside else blocks. Also change if statements to the following format: 

```if (myCondition) {``` 

(space after `if` and space before open bracket)

File: GPSSetup/clients/GPSSetup/controllers/page/CameraGPS/onStart.js [hunk 33]
```diff
@@ -0,0 +1,197 @@
...
+
+        }    
+
+        if(page.data.camera.imageTrack.topRightImage){
+            var x = page.data.camera.imageTrack.topRightImage.x;
+            var y = page.data.camera.imageTrack.topRightImage.y;
+            page.data.CanvasState.addShape(new Shape(x,y,20,20, 'rgba(0,255,0, 0.6)'));
+        } else {
+            page.data.CanvasState.addShape(new Shape(80,20,20,20, 'rgba(0,255,0, 0.6)'));
+        }
```
Comment (namirfa) [inline GPSSetup/clients/GPSSetup/controllers/page/CameraGPS/onStart.js:173]: Ditto for if styling

File: GPSSetup/clients/GPSSetup/controllers/page/CameraGPS/onStart.js [hunk 34]
```diff
@@ -0,0 +1,197 @@
...
+
+        } else {
+            page.data.CanvasState.moveShape(3, 80,20);
+        }
+
+
+    }
+
+    function drawShape(){
+        //. then if there are existing plots use them instead
```
Comment (namirfa) [inline GPSSetup/clients/GPSSetup/controllers/page/CameraGPS/onStart.js:141]:
```suggestion
        // then if there are existing plots use them instead
```

File: GPSSetup/clients/GPSSetup/controllers/page/CameraGPS/widgets/Button3523/onClick.js [hunk 35]
```diff
@@ -0,0 +1,12 @@
+
+
```
Comment (namirfa) [inline GPSSetup/clients/GPSSetup/controllers/page/CameraGPS/widgets/Button3523/onClick.js:2]: Remove initial empty lines. Also same question about indentation.

File: GPSSetup/clients/GPSSetup/controllers/page/CameraGPS/widgets/Button3524/onClick.js [hunk 36]
```diff
@@ -0,0 +1,4 @@
+
+
+
```
Comment (namirfa) [inline GPSSetup/clients/GPSSetup/controllers/page/CameraGPS/widgets/Button3524/onClick.js:3]: Remove initial empty lines. Also same question about indentation.

File: GPSSetup/clients/GPSSetup/controllers/page/CameraGPS/widgets/Button3525/onClick.js [hunk 37]
```diff
@@ -0,0 +1 @@
+	client.returnToCallingPage();
```
Comment (namirfa) [inline GPSSetup/clients/GPSSetup/controllers/page/CameraGPS/widgets/Button3525/onClick.js:1]: Same question about indentation.

File: GPSSetup/clients/GPSSetup/controllers/page/Start/onStart.js [hunk 38]
```diff
@@ -0,0 +1,55 @@
+
```
Comment (namirfa) [inline GPSSetup/clients/GPSSetup/controllers/page/Start/onStart.js:1]: Remove initial empty line. Also same question about indentation.

File: GPSSetup/clients/GPSSetup/controllers/page/Start/onStart.js [hunk 39]
```diff
@@ -0,0 +1,55 @@
...
+    map.addListener("click", (mapsMouseEvent) => {
+        // Close the current InfoWindow.
+        infoWindow.close();
+        // Create a new InfoWindow.
+        infoWindow = new google.maps.InfoWindow({
+            position: mapsMouseEvent.latLng,
+        });
+        //infoWindow.setContent(
+        //    JSON.stringify(mapsMouseEvent.latLng.toJSON(), null, 2)
+        //);
```
Comment (namirfa) [inline GPSSetup/clients/GPSSetup/controllers/page/Start/onStart.js:29]: Remove commented code

File: GPSSetup/clients/GPSSetup/controllers/page/Start/onStart.js [hunk 40]
```diff
@@ -0,0 +1,55 @@
...
+
+        var mmjson = mapsMouseEvent.latLng.toJSON();
+
```
Comment (namirfa) [inline GPSSetup/clients/GPSSetup/controllers/page/Start/onStart.js:32]: Remove extra empty line

File: GPSSetup/clients/GPSSetup/controllers/page/Start/onStart.js [hunk 41]
```diff
@@ -0,0 +1,55 @@
...
+                console.log(args);
+                console.log("SUCCESS: " + JSON.stringify(response));
+                page.data.refreshCamerasTable();
+            });
+        } else {
+            infoWindow.setContent(
+                "nothing selected - nothing added"
+            );
+        }
+        //infoWindow.open(map);
```
Comment (namirfa) [inline GPSSetup/clients/GPSSetup/controllers/page/Start/onStart.js:54]: Remove commented code

File: GPSSetup/clients/GPSSetup/controllers/page/Start/widgets/DataTable492/onButtonClickCallback.js [hunk 42]
```diff
@@ -0,0 +1,23 @@
+    console.log("Extra: " + extra);
```
Comment (namirfa) [inline GPSSetup/clients/GPSSetup/controllers/page/Start/widgets/DataTable492/onButtonClickCallback.js:1]: Same question about indentation

File: GPSSetup/clients/GPSSetup/controllers/page/Start/widgets/DataTable492/onButtonClickCallback.js [hunk 43]
```diff
@@ -0,0 +1,23 @@
...
+
+                client.deleteOne("cameras",cameraId,function(response){
+                    page.data.refreshCamerasTable();
+                });
+
+            }
+        });  
+
+    }
+
```
Comment (namirfa) [inline GPSSetup/clients/GPSSetup/controllers/page/Start/widgets/DataTable492/onButtonClickCallback.js:23]: Remove unnecessary empty lines

File: GPSSetup/clients/GPSSetup/controllers/sidebar/widgets/MenuButton4/onClick.js [hunk 44]
```diff
@@ -0,0 +1 @@
+    client.goToPage("Start");
```
Comment (namirfa) [inline GPSSetup/clients/GPSSetup/controllers/sidebar/widgets/MenuButton4/onClick.js:1]: Same question about indentation

File: GPSSetup/procedures/submitCamera.vail [hunk 45]
```diff
@@ -0,0 +1,4 @@
+PROCEDURE submitCamera(obj Object)
+
+UPSERT cameras(obj)
+
```
Comment (namirfa) [inline GPSSetup/procedures/submitCamera.vail:4]: Unnecessary empty line

File: GPSSetup/procedures/transformCameraEvent.vail [hunk 46]
```diff
@@ -0,0 +1,26 @@
+PROCEDURE transformCameraEvent(event)
+
+var calibrationCoordinates = {imageCoordinates:[{x:event.cameras.imageTrack.topLeftImage.x, y:event.cameras.imageTrack.topLeftImage.y}, {x:event.cameras.imageTrack.topRightImage.x, y:event.cameras.imageTrack.topRightImage.y}, {x:event.cameras.imageTrack.bottomRightImage.x, y:event.cameras.imageTrack.bottomRightImage.y}, {x:event.cameras.imageTrack.bottomLeftImage.x, y:event.cameras.imageTrack.bottomLeftImage.y}], convertedCoordinates: [{x:event.cameras.gpsTrack.topLeftGeo.lng, y:event.cameras.gpsTrack.topLeftGeo.lat}, {x:event.cameras.gpsTrack.topRightGeo.lng, y:event.cameras.gpsTrack.topRightGeo.lat}, {x:event.cameras.gpsTrack.bottomRightGeo.lng, y:event.cameras.gpsTrack.bottomRightGeo.lat}, {x:event.cameras.gpsTrack.bottomLeftGeo.lng, y:event.cameras.gpsTrack.bottomLeftGeo.lat}]}
+
+var results = []
+
+for each detection in event.detections {
+    if(detection.label == "person"){
```
Comment (namirfa) [inline GPSSetup/procedures/transformCameraEvent.vail:8]: Same comment about `if` codestyling

**Review by namirfa — CHANGES_REQUESTED** [review]
Looks good, most comments are related to code-style.

I tried testing this manually and I think it works as expected, though I'm not confident that I tested it thoroughly enough. It would be great if you could show me a demonstration of how it works so that I can more effectively test edge cases. That will also help me give notes for clarifying the Project Description.

**Review by pburma — COMMENTED** [review]
accepted changes

**Review by steve-langley — COMMENTED** [review]
Some observations...

It doesn't look like the icons/acousticIcon.png or icons/cameraIcon.png Documents are actually checked into the GPSSetup project (at least I couldn't find them)

"ConvertCoordinates" is misspelled a couple of times in the Project Description (once as "CovertCoordinates" and once as "ConvertCoordinated")

Don't take the comments below too seriously - just impressions from someone who really doesn't know what this client is doing...

Is there a reason why there is a DropList on the Start page with only a single item in the list? ("Camera")

Could probably use some more comments. For example, some explanation as to what the CanvasState class in "Custom Code" does.


Maybe add a comment in the "OnClick" event for the Load Image button that explains that "#imageLoader" is found in the StaticHtml named "StaticHtml897" and this is a trick to upload an image file. The event listeners to handle this are in the CameraGPS Page's "On Start" event.

Is the sidebar with "Start" really necessary? The "Back" button on the CameraGPS Page does the same thing.

FWIW "callibration-canvas" should be spelled "calibration-canvas"

**Comment by zhangevan** [comment]
> The GPS and XY coordinates are saved to the cameras type. Once saved to the cameras type a CachedEnriched event can be used to load those values into each memory and then passed into a ConvertCoordinated activity pattern as a downstream event. An example application in the App Builder will look like this:
> 
> EventStream(objrec payload) --> SplitBy CameraId --> CachedEnrich cameras --> Transformation --> ConvertCoordinates --> TrackMotion

Should we create this App and include it in the contribution to further demonstrate the usage of those coordinates?

**Review by rblumer4 — APPROVED** [review]
May still be missing an icon file.
But this isn't going to be added to the Import dialog, so that can be added in a later PR.
